### ISSUE TITLE
More numpy-specific general type issues (_ArrayLikeFloat_co, _NestedSequence)

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#6523]
I hope this doesn't turn out to be a support question, since my code below behaves so differently across `pyright` versions that I am starting to wonder what is expected behavior and what is not :) I checked release notes of, in particular, 1.1.333 and 1.1.337 and did not see any enhancements that might (intend to) change typechecking of the code below.

**Describe the bug**
I cannot pass lists of tuples of floats to `numpy` any longer. This started in 1.1.337, but it seem parts of it started in 1.1.333 when I isolate the issue.

**Code or Screenshots**
```python
# Example
import numpy as np

def example(arg: list[tuple[float, float]]) -> None:
    np.histogram2d(x=1, y=1, range=arg)


# Minimized
from numpy._typing._array_like import (
    _ArrayLikeFloat_co,  # pyright: ignore[reportPrivateUsage]
)

def inner1(arg: _ArrayLikeFloat_co) -> None:
    print(arg)

def outer1(arg: list[tuple[float, float]]) -> None:
    inner1(arg)


# Digging deeper, list[tuple[float, float]] is not a _NestedSequence[float]?
from numpy._typing._nested_sequence import _NestedSequence

def inner2(arg: _NestedSequence[float]) -> None:
    print(arg)

def outer2a(arg: list[tuple[float, float]]) -> None:
    inner2(arg)


# But list[list[float]] is!
def outer2b(arg: list[list[float]]) -> None:
    inner2(arg)
```

pyright via pyright-python.

v1.1.332:
```
0 errors, 0 warnings, 0 informations 
```

v1.1.333-336:
```
C:\Build\project\bug.py:6:5 - error: No overloads for "histogram2d" match the provided arguments (reportGeneralTypeIssues)
  C:\Build\project\bug.py:20:12 - error: Argument of type "list[tuple[float, float]]" cannot be assigned to parameter "arg" of type "_ArrayLikeFloat_co" in function "inner1"       
    Type "list[tuple[float, float]]" cannot be assigned to type "_ArrayLikeFloat_co"
      "list[tuple[float, float]]" is incompatible with protocol "_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]"
        "__array__" is not present
      "list[tuple[float, float]]" is incompatible with protocol "_NestedSequence[_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]]"
        "__getitem__" is an incompatible type
          No overloaded function matches type "(index: int, /) -> (_T_co@_NestedSequence | _NestedSequence[_T_co@_NestedSequence])"
        "__iter__" is an incompatible type
          Type "() -> Iterator[tuple[float, float]]" cannot be assigned to type "() -> Iterator[_T_co@_NestedSequence | _NestedSequence[_T_co@_NestedSequence]]" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

v1.1.337:
```
C:\Build\project\bug.py:6:5 - error: No overloads for "histogram2d" match the provided arguments (reportGeneralTypeIssues)
  C:\Build\project\bug.py:6:36 - error: Argument of type "list[tuple[float, float]]" cannot be assigned to parameter "range" of type "_ArrayLikeFloat_co | None" in function "histogram2d"
    Type "list[tuple[float, float]]" cannot be assigned to type "_ArrayLikeFloat_co | None"
      "list[tuple[float, float]]" is incompatible with "None"
      "list[tuple[float, float]]" is incompatible with protocol "_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]"
        "__array__" is not present
      "list[tuple[float, float]]" is incompatible with protocol "_NestedSequence[_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]]"
        "__getitem__" is an incompatible type
          No overloaded function matches type "(index: int, /) -> (_T_co@_NestedSequence | _NestedSequence[_T_co@_NestedSequence])"
        "__iter__" is an incompatible type
    ... (reportGeneralTypeIssues)
  C:\Build\project\bug.py:20:12 - error: Argument of type "list[tuple[float, float]]" cannot be assigned to parameter "arg" of type "_ArrayLikeFloat_co" in function "inner1"       
    Type "list[tuple[float, float]]" cannot be assigned to type "_ArrayLikeFloat_co"
      "list[tuple[float, float]]" is incompatible with protocol "_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]"
        "__array__" is not present
      "list[tuple[float, float]]" is incompatible with protocol "_NestedSequence[_SupportsArray[dtype[bool_ | integer[Any] | floating[Any]]]]"
        "__getitem__" is an incompatible type
          No overloaded function matches type "(index: int, /) -> (_T_co@_NestedSequence | _NestedSequence[_T_co@_NestedSequence])"
        "__iter__" is an incompatible type
          Type "() -> Iterator[tuple[float, float]]" cannot be assigned to type "() -> Iterator[_T_co@_NestedSequence | _NestedSequence[_T_co@_NestedSequence]]"
    ... (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations
```

So I guess that similar as in https://github.com/microsoft/pyright/issues/6511 (in 1.1.337 but not 1.1.336), the recursion depth may play a role here, which is why I am seeing these issues appear in my code base in 1.1.337. (Another point which makes me think that is that by adding the above `bug.py` to my code base, I also see errors using 1.1.336 which I do not without `bug.py`.)

I just don't know if these errors I am seeing are false-positives introduced, or false-negatives fixed in 1.1.333. Since `np.array([(1., 1.)])` works, I am inclined to think it's the former, but my knowledge of the whole contra/covariance stuff regarding different sequence/container types is rather limited and I may easily be wrong.

**Comment by bersbersbers** [comment]
Here's another one that typechecks in 1.1.336 but not in 1.1.337:

```python
import numpy as np
import numpy.typing as npt

def fun1(x: npt.NDArray[np.str_]) -> None:
    np.ndenumerate(x)

def fun2(x: list[list[str]]) -> None:
    np.ndenumerate(x)

def fun12(x: npt.NDArray[np.str_] | list[list[str]]) -> None:
    np.ndenumerate(x)
```

This one I think I understand, since there are several overloads of `ndenumerate.__new__` which work individually but maybe not  for the union-typed argument. Still, since that one worked in 1.1.336, it may be an unintended change as well?

**Comment by erictraut** [comment]
Is your source in a public repo? If so, it would be useful if you'd add it to [mypy_primer](https://github.com/hauntsaninja/mypy_primer). This is a tool that we run on every pyright commit to catch regressions. We'd need to make sure that the numpy dependency is installed as part of the mypy_primer profile.

**Comment by bersbersbers** [comment]
> Is your source in a public repo?

Unfortunately not :(

**Comment by erictraut** [comment]
My previous diagnosis addressed only part of the issue. Pyright contains a "protocol matching cache" used to skip redundant type compatibility checks with protocols. This is especially effective for complex and deeply-nested protocols like those used in numpy such as `_NestedSequence`. This cache had a subtle bug, which resulted in the behavior you're seeing.

Writing tests for cases like this is very difficult. It would be great if we could get a real-world public source base into the mypy_primer tool so we could catch future regressions like this. I understand that your code isn't public, so we'll need to look for other similar code bases that are.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.338, which I just published. It will also be included in a future release of pylance.
